sherlockml/clients/report.py

```diff
@@ -0,0 +1,174 @@
+# Copyright 2018 ASI Data Science
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import namedtuple
+
+from marshmallow import Schema, fields, post_load, pre_load
+
+from sherlockml.clients.base import BaseClient
+
+Report = namedtuple(
+    "Report", ["created_at", "name", "id", "description", "active_version"]
+)
+
+VersionedReport = namedtuple(
+    "VersionedReport",
+    [
+        "created_at",
+        "name",
+        "id",
+        "description",
+        "active_version_id",
+        "versions",
+    ],
+)
+
+ReportVersion = namedtuple(
+    "ReportVersion",
+    [
+        "created_at",
+        "author_id",
+        "report_path",
+        "report_key",
+        "report_bucket",
+        "notebook_path",
+        "id",
+        "report_id",
+    ],
+)
+
+
+class ReportVersionSchema(Schema):
+    def __init__(self, report_id=None, *args, **kwargs):
+        super(ReportVersionSchema, self).__init__(*args, **kwargs)
+        if report_id is not None:
+            self.context["report_id"] = report_id
```

Comment:
I'm not crazy about this use of the schema context to pass around values like this - it introduces a lot of coupling between the schemas. Why is it necessary to have the report ID in the version model?

Reply:
This is to get around an inconsistency in what Tavern returns:

- GET report/<report_id>/versions returns a report, with a list of versions, which *don't* contain the report_id
- POST report/<report_id>/version returns just a version, which *also* contains the report_id

I think I would prefer to have the python representation of a Report Version be consistent in what it contains, hence the use of context